Py script to generate a .dot graph of rules/groups and their edges

dvmarcilio opened the pull request:
The script generates a .dot file used to generate a graph (.dot) for Polyglot's rules and edges through [graphviz](https://graphviz.org/).

I added a graph image to the README of the pre-built Java cleanup rules.
I also added a section with instructions on how to run the script and its requirements.

ketkarameya commented on polyglot/piranha/rules_dot_graph.py:
```diff
@@ -0,0 +1,177 @@
+        if 'groups' in rule_toml:
+            for group_name in rule_toml['groups']:
+                group_name = sanitize_name(group_name)
+                if group_name != 'Cleanup_Rule':
```
i think there is no `_`

dvmarcilio replied:
You're right. In `rules.toml`, and probably in the rust code, it's `Cleanup Rule`.
For Graphviz, we are converting all blank spaces to `_`. The conversion makes our lives easier when defining nodes and edges. I'm not sure Graphviz plays well with spaces in names.

https://github.com/uber/piranha/pull/262/files/5b26d2d4c3dae196cf76376509182a11e7e9cd1f#diff-62d9897c4603bf7c92c8ea08ac5a69b1cb508aa8f9cf262bac0e3d83649fc425R48-R51

ketkarameya replied:
I see. I see. Could be a good naming convention to add .

ketkarameya commented on polyglot/piranha/visualize_rules_graph.py:
```diff
@@ -0,0 +1,234 @@
+                outgoing_edges_by_node[from_node] = [edge]
+
+
+def group_rule_names_label(rule_names: 'list[str]') -> 'list[str]':
```
What are we grouping here? 
When you say `group_<something>`  I guess the function should return a `dict` or a `list of list`.

ketkarameya replied:
I guess this is the _"group"_ from our `Rule`. :|

dvmarcilio replied:
I changed the name to make it more clear.

ketkarameya commented on polyglot/piranha/visualize_rules_graph.py:
```diff
@@ -0,0 +1,234 @@
+def inline_name(rule_name: str) -> str:
+    """Should be called for rules under a group."""
+    if rule_name in cleanup_rules:
+        return f'{rule_name} (Cleanup Rule)'
```
Why are we appending `(Cleanup Rule)` ?

ketkarameya commented on polyglot/piranha/visualize_rules_graph.py:
```diff
@@ -0,0 +1,234 @@
+        if 'groups' in rule_toml:
+            rule_groups = rule_toml['groups']
+            collect_node_for_group_rule(rule_name, rule_groups)
+        else:  # rule without group
```
why would node without a `group` be orphan node? Shouldn't it be when it is a node without an edge?

ketkarameya commented on polyglot/piranha/visualize_rules_graph.py:
```diff
@@ -0,0 +1,234 @@
+
+
+def all_non_orphan_rule_names() -> 'list[str]':
+    node_names_with_only_outgoing_edges = [
```
is this the [set difference](https://www.geeksforgeeks.org/python-set-difference/) between the keys of two dicts?

dvmarcilio replied:
Nice catch. I changed it to use the set difference operator

ketkarameya commented on polyglot/piranha/visualize_rules_graph.py:
```diff
@@ -0,0 +1,234 @@
+
+
+def generate_orphan_nodes():
+    # orphan_nodes don't have any group, not even 'Cleanup_Rule'
```
Same comment as above. Are we using the term `orphan node` in a non standard way?

ketkarameya reviewed (changes requested):
Left a few comments.

dvmarcilio commented:
Thanks for the careful review. It made me realize that the script was not that clear. 
I changed some names and added more comments. It should be better now.

ketkarameya commented on polyglot/piranha/visualize_rules_graph.py:
```diff
@@ -0,0 +1,257 @@
+                rules_by_group_dict[group_name].append(rule_name)
+            else:
+                rules_by_group_dict[group_name] = [rule_name]
+        else:  # we don't want to group 'Cleanup Rule's under the same graphviz shape
```
```suggestion
        else:  # we don't want to group `Cleanup Rule`s under the same graphviz shape
```

ketkarameya reviewed (approved):
small nit. please merge the PR after addressing it.
